paz/backend/boxes.py

```diff
@@ -150,231 +120,434 @@ def key_is_pressed(key="q", time=20):
-        selections = jp.concatenate(
-            (boxes[selected_indices], scores[selected_indices]), axis=1
+def create_outputs(num_classes, decoded_boxes, class_predictions, NMS_thresh, confidence_thresh, top_k):
+    """Create outputs for each class after NMS."""
+    outputs = []
+    for i in range(num_classes):
+        output = process_class_output(
+            i, decoded_boxes, class_predictions, NMS_thresh, confidence_thresh, top_k
         )
-        output = output.at[class_arg, :count, :].set(selections)
-    return output
+        outputs.append(output)
+    return outputs
 
 
-def compute_iou(box, boxes):
-    """Calculates the intersection over union between 'box' and all 'boxes'.
-    Both `box` and `boxes` are in corner coordinates.
+def NMS_per_class(box_data, NMS_thresh=0.45, confidence_thresh=0.01, top_k=200):
+    """Apply NMS per class to decoded box data.
 
-    # Arguments
-        box: JAX array with length at least of 4.
-        boxes: JAX array with shape `(num_boxes, 4)`.
+    Args:
+        box_data (array): Decoded box predictions with class scores
+        NMS_thresh (float): IoU threshold for NMS
+        confidence_thresh (float): Minimum confidence score
+        top_k (int): Maximum detections per class
 
-    # Returns
-        JAX array of shape `(num_boxes, 1)`.
+    Returns:
+        array: Selected boxes and scores per class
     """
+    decoded_boxes, class_predictions = decode_box_data(box_data)
+    num_classes = class_predictions.shape[1]
+    outputs = create_outputs(
+        num_classes, decoded_boxes, class_predictions, NMS_thresh, confidence_thresh, top_k
+    )
+    return jp.stack(outputs, axis=0)
 
+
+def get_boxes_coordinates_to_compute_IoU(box, boxes):
+    """Get coordinates for IoU computation between a box and others."""
     x_min_A, y_min_A, x_max_A, y_max_A = box[:4]
     x_min_B, y_min_B = boxes[:, 0], boxes[:, 1]
     x_max_B, y_max_B = boxes[:, 2], boxes[:, 3]
-    # calculating the intersection
+    return (
+        x_min_A,
+        y_min_A,
+        x_max_A,
+        y_max_A,
+        x_min_B,
+        y_min_B,
+        x_max_B,
+        y_max_B,
+    )
+
+
+def get_inner_coordinates(x_min_A, y_min_A, x_max_A, y_max_A, x_min_B, y_min_B, x_max_B, y_max_B):
+    """Compute inner coordinates for intersection area."""
     inner_x_min = jp.maximum(x_min_B, x_min_A)
     inner_y_min = jp.maximum(y_min_B, y_min_A)
     inner_x_max = jp.minimum(x_max_B, x_max_A)
     inner_y_max = jp.minimum(y_max_B, y_max_A)
+    return inner_x_min, inner_y_min, inner_x_max, inner_y_max
+
+
+def calculate_intersection(inner_x_min, inner_y_min, inner_x_max, inner_y_max):
+    """Calculate intersection area between two boxes."""
     inner_w = jp.maximum((inner_x_max - inner_x_min), 0)
     inner_h = jp.maximum((inner_y_max - inner_y_min), 0)
-    intersection_area = inner_w * inner_h
-    # calculating the union
+    return inner_w * inner_h
+
+
+def calculate_union(
```

Comment:
Most  of the time we say "compute" rather than "calculate" since compute is smaller.